internal/throttler/throttler.go

```diff
@@ -0,0 +1,115 @@
+//go:generate mockgen -source throttler.go -destination ../mocks/mock_throttler.go -package mocks
+
+package throttler
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/openfga/openfga/internal/build"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/openfga/openfga/pkg/telemetry"
+)
+
+var (
+	throttlingDelayMsHistogram = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace:                       build.ProjectName,
+		Name:                            "throttling_delay_ms",
+		Help:                            "Time spent waiting for dispatch throttling resolver",
+		Buckets:                         []float64{1, 3, 5, 10, 25, 50, 100, 1000, 5000}, // Milliseconds. Upper bound is config.UpstreamTimeout.
+		NativeHistogramBucketFactor:     1.1,
+		NativeHistogramMaxBucketNumber:  100,
+		NativeHistogramMinResetDuration: time.Hour,
+	}, []string{"grpc_service", "grpc_method", "throttler_name"})
+)
+
+type Throttler interface {
+	Close()
+	Throttle(context.Context)
+}
+
+type noopThrottler struct{}
+
+var _ Throttler = (*noopThrottler)(nil)
+
+func (r *noopThrottler) Throttle(ctx context.Context) {
+}
+
+func (r *noopThrottler) Close() {
+}
+
+func NewNoopThrottler() Throttler { return &noopThrottler{} }
+
+// constantRateThrottler implements a throttling mechanism that can be used to control the rate of recursive resource consumption.
+// Throttling will release the goroutines from the throttlingQueue based on the configured ticker.
+type constantRateThrottler struct {
+	name            string
+	ticker          *time.Ticker
+	throttlingQueue chan struct{}
+	done            chan struct{}
+}
+
+// NewConstantRateThrottler constructs a constantRateThrottler which can be used to control the rate of recursive resource consumption.
+func NewConstantRateThrottler(frequency time.Duration, metricLabel string) Throttler {
```

Comment:
Ask: Can you godoc whether this throttler object is supposed to be instantiated once at the server level and not once per request?

Thought: Right now all the check resolvers seem to be designed at the server level... i'm worried for the future, when we add options to tune caching per-request, the cached Check Resolver will have to be instantiated at the request level, this will throw off the entire resolver onion?